### PR TITLE
WebLab: version history, autosave, and project template support

### PR DESCRIPTION
**Pull request by cpirich**
- studioApp.js: broke out version history UI init code into its own function `initVersionHistoryUI()` so that it can be called be weblab, which doesn't call the full `studioApp.init()` function
- Allowed an app to specify `config.versionHistoryInInstructionsHeader` to indicate that it wants to show a version history button in the instructions header. This is used by weblab because it doesn't have the `codeModeHeaders` toolbar from `CodeWorkspace.jsx`
- Changed `initApp.js` to treat weblab like applab and gamelab with respect to showing "version history" instead of "clear puzzle"
- Weblab now registers a callback to notify the dashboard that the project has changed using a new brambleHost `onProjectChanged` method.
- brambleHost now connects with `fileChange`, `fileDelete`, `fileRename`, and `folderRename` events in the bramble file system so it can notify the upstream app (weblab) that the project has changed.

### CODE REVIEW

**Comment by cpirich** [comment]
@jeremydstone

**Comment by jeremydstone** [comment]
LGTM

**Comment by cpirich** [comment]
Additional changes (with a merge of latest from staging in the hopes that it will resolve UI test failures that seemed unrelated to my changes):
- make `handleClearPuzzle` return a `Promise` so that if clearing the editor is an async operation (as is the case with weblab's use of bramble and its IndexedDB file system), the subsequent "save" operation will occur after the async "clear" is complete.
- update dashboard code for weblab to support `project_template_level` both in levelbuilder and at runtime by ensuring that `startSources` will use the version from the project template level

**Comment by cpirich** [comment]
@jeremydstone

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -811,8 +813,15 @@ StudioApp.prototype.handleClearPuzzle = function (config) {
     annotationList.clearRuntimeAnnotations();
   }
   if (config.afterClearPuzzle) {
-    config.afterClearPuzzle();
+    promise = config.afterClearPuzzle(config);
+  }
+  if (!promise) {
```
Comment (jeremydstone) [inline apps/src/StudioApp.js:818]: I feel like it starts heading into spaghetti-code territory when afterClearPuzzle might return void or might return a promise. It looks like there are 5 implementations of afterClearPuzzle (applab.js, GameLab.js, studio.js, blockly/js/gamelab.js, studio.js) and my preference would be to add a one liner in each to return a promise. Could have a wrapper in the base like CreateInstaPromise() to make it easy to read.

Comment (cpirich) [reply apps/src/StudioApp.js:818]: Agreed, but many apps don't even implement `afterClearPuzzle()`, so I already need to handle this case for maze, turtle, eval, bounce, etc. I don't want to make all of them implement it just to return the "insta promise" (I like that name).
And unfortunately, `StudioApp` is not a true base class, so it's not like the apps will automatically get this behavior unless I implement it here in `handleClearPuzzle()`

**Comment by cpirich** [comment]
Test failures are the same as those appearing in other PRs on staging (not related to these changes), so merging...

**Pull request merged**